core/src/main/scala/org/apache/spark/sql/delta/commands/RestoreTableCommand.scala

```diff
@@ -0,0 +1,184 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.commands
+
+import java.sql.Timestamp
+import org.apache.spark.sql.delta._
+import org.apache.spark.sql.delta.actions.{AddFile, RemoveFile}
+import org.apache.spark.sql.delta.sources.DeltaSQLConf._
+import org.apache.spark.sql.delta.util.DeltaFileOperations.absolutePath
+import org.apache.spark.sql.execution.command.LeafRunnableCommand
+import org.apache.spark.sql.{Dataset, Row, SparkSession}
+import org.apache.spark.sql.internal.SQLConf.IGNORE_MISSING_FILES
+import org.apache.spark.util.SerializableConfiguration
+
+/**
+ * Perform restore of delta table to a specified version or timestamp
+ *
+ * Algorithm:
+ *  1) Read the latest snapshot of the table.
+ *  2) Read snapshot for version or timestamp to restore
+ *  3) Compute files available in snapshot for restoring (files were removed by some commit)
+ *  but missed in the latest. Add these files into commit as AddFile action.
+ *  4) Compute files available in the latest snapshot (files were added after version to restore)
+ *  but missed in the snapshot to restore. Add these files into commit as RemoveFile action.
+ *  5) If SQLConf.IGNORE_MISSING_FILES option is false (default value) check availability of AddFile
+ *  in file system.
+ *  6) Commit metadata, Protocol, all RemoveFile and AddFile actions
+ *  into delta log using `commitLarge`.
+ *  7) If table was modified in parallel then ignore restore and raise exception.
+ */
+case class RestoreTableCommand(
+  deltaLog: DeltaLog,
+  version: Option[Long],
+  timestamp: Option[Timestamp]
+) extends LeafRunnableCommand with DeltaCommand {
+
+  override def run(spark: SparkSession): Seq[Row] = {
+    recordDeltaOperation(deltaLog, "delta.restore") {
+
+      require(version.isEmpty ^ timestamp.isEmpty,
+        "Either the version or timestamp should be provided for restore")
+      val parallelism = restoreParallelism(spark)
+      val latestSnapshot = deltaLog.update()
+      val versionToRestore = version.getOrElse(
+        deltaLog
+          .history
+          .getActiveCommitAtTime(timestamp.get, canReturnLastCommit = true)
+          .version
+      )
+
+      require(versionToRestore < latestSnapshot.version,
+        s"Version to restore ($versionToRestore) should be less then " +
+        s"last available version (${latestSnapshot.version})")
+
+      val snapshotToRestore = deltaLog.getSnapshotAt(versionToRestore)
+      val latestSnapshotFiles = latestSnapshot.allFiles
+      val snapshotToRestoreFiles = snapshotToRestore.allFiles
+
+      import spark.implicits._
+      import collection.JavaConverters._
+
+      val filesToAdd = snapshotToRestoreFiles
+        .join(
+          latestSnapshotFiles,
+          snapshotToRestoreFiles("path") ===  latestSnapshotFiles("path"),
+          "left_anti")
+        .as[AddFile]
+        .map(_.copy(dataChange = true))
+        .repartition(parallelism)
```

Comment:
why do you need to repartitoin to a specific parallelism. why not just rely on the join using the standard spark.sql.shuffle.partitions to decide this?